### PR TITLE
Feature ETP-3752: Fix display logic

### PR DESCRIPTION
**Pull request by leandroallemandismf**

### CODE REVIEW

File: packages/MainUI/utils/expressions.ts [hunk 1]
```diff
@@ -216,6 +220,21 @@ export const createEvaluationContext = (options: SmartContextOptions) => {
...
       // removing the null->'' conversion that getValue provided. The Proxy must handle it.
 
+      // Dev-mode warning: surface @field_name@ tokens that cannot be resolved so
+      // developers can spot misconfigured Application Dictionary expressions early.
+      // warnedRefs deduplicates so each missing key is logged at most once per context.
+      if (
+        process.env.NODE_ENV !== "production" &&
+        /^[a-zA-Z_]\w*$/.test(prop) &&
+        !["then", "toJSON", "toString", "valueOf", "constructor", "Symbol"].includes(prop) &&
+        !warnedRefs.has(prop)
```
Comment (etendobot) [inline packages/MainUI/utils/expressions.ts:230]: **Suggestion**

Remove 'Symbol' from the exclusion array as it is already handled by the type check at the beginning of the Proxy trap.

```typescript
        !["then", "toJSON", "toString", "valueOf", "constructor"].includes(prop) &&
```

**Rationale**: Since the 'get' trap already returns early if the property is not a string (line 204), Symbol properties will never reach this logic. Removing 'Symbol' from the exclusion list is a minor cleanup that reduces redundancy.

File: packages/MainUI/utils/__tests__/expressions/dependencies.test.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    it("extracts field deps but excludes session deps in mixed expression", () => {
+      const deps = extractDependenciesFromExpression("@fieldA@='Y' & @#AD_Org_ID@='1'");
+      expect(deps).toEqual(["fieldA"]);
+    });
+  });
+
+  describe("DB column → HQL name mapping via field metadata", () => {
+    const fields: Record<string, Field> = {
+      trxtypeDisplayLogic: {
+        hqlName: "trxtypeDisplayLogic",
```
Comment (etendobot) [inline packages/MainUI/utils/__tests__/expressions/dependencies.test.ts:66]: **Suggestion**

Consider using 'Partial<Field>' or a dedicated mock utility instead of 'as any' to improve type safety in tests.

```typescript
      trxtypeDisplayLogic: {
        hqlName: "trxtypeDisplayLogic",
        columnName: "TRXTYPE_DISPLAY_LOGIC",
        column: { dBColumnName: "TRXTYPE_DISPLAY_LOGIC" },
      } as Partial<Field> as Field,
```

**Rationale**: Using 'as any' bypasses type checking. While acceptable in tests for complex objects, using 'Partial<Field>' or a more specific type assertion provides better safety and documentation of which properties are actually required for the test.

File: packages/MainUI/__tests__/utils/parseDynamicExpression.test.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+ * (the "License"), you may not use this file except in compliance with
+ * the License.
+ * You may obtain a copy of the License at
+ * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
+ * Software distributed under the License is distributed on an
+ * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
+ * implied. See the License for the specific language governing rights
+ * and limitations under the License.
+ * All portions are Copyright © 2025 FUTIT SERVICES, S.L
+ * All Rights Reserved.
```
Comment (etendobot) [inline packages/MainUI/__tests__/utils/parseDynamicExpression.test.ts:13]: **Suggestion**

Use '(C)' instead of '©' in the copyright line of the license header.

```typescript
 * All portions are Copyright (C) 2025 FUTIT SERVICES, S.L
```

**Rationale**: While technically equivalent, the Etendo standard license header typically uses '(C)' instead of the '©' symbol. Maintaining this consistency ensures compatibility with automated tools and follows established patterns.

File: packages/MainUI/components/Form/FormView/selectors/BaseSelector.tsx [hunk 4]
```diff
@@ -114,10 +114,14 @@ export const compileExpression = (expression: string) => {
...
+    //
+    // Normalize the expression result to a proper JS boolean so that bare @field@ references
+    // (which yield Etendo 'Y'/'N' strings) behave correctly: 'Y' → true, 'N' → false.
+    // Comparison expressions (==, !=, &&, ||) already return JS booleans and pass through unchanged.
     const compiled = new Function(
       "context",
       "currentValues",
-      `${securityShim} ${obShim} return ${parseDynamicExpression(expression)};`
+      `${securityShim} ${obShim} var __r = (${parseDynamicExpression(expression)}); return __r === 'N' ? false : __r === 'Y' ? true : Boolean(__r);`
     );
```
Comment (etendobot) [inline packages/MainUI/components/Form/FormView/selectors/BaseSelector.tsx:126]: **Suggestion**

Replace 'var' with 'const' for the temporary result variable '__r'.

```typescript
      `${securityShim} ${obShim} const __r = (${parseDynamicExpression(expression)}); return __r === 'N' ? false : __r === 'Y' ? true : Boolean(__r);`
```

**Rationale**: In modern Etendo UI development using TypeScript and React, 'const' is the preferred keyword for variables that are not reassigned. It provides better scoping and aligns with the coding standards used throughout the MainUI package.

**Review by etendobot — APPROVED** [review]

File: packages/MainUI/utils/__tests__/expressions/dependencies.test.ts (hunk 2 above)
Comment (etendobot) [inline packages/MainUI/utils/__tests__/expressions/dependencies.test.ts:66]: **Suggestion**

Consider avoiding 'as any' by using 'as unknown as Field' or defining a mock helper for Field objects to maintain type safety.

```typescript
      trxtypeDisplayLogic: {
        hqlName: "trxtypeDisplayLogic",
        columnName: "TRXTYPE_DISPLAY_LOGIC",
        column: { dBColumnName: "TRXTYPE_DISPLAY_LOGIC" },
      } as unknown as Field,
```

**Rationale**: Using 'as any' bypasses TypeScript's type safety. While acceptable in tests for complex objects, using a more specific cast or a mock helper is preferred to ensure the test objects remain consistent with the 'Field' interface.

File: packages/MainUI/__tests__/utils/parseDynamicExpression.test.ts (hunk 3 above)
Comment (etendobot) [inline packages/MainUI/__tests__/utils/parseDynamicExpression.test.ts:13]: **Suggestion**

Update the copyright line to use '(C)' and the standard year range if this module follows the core versioning.

```typescript
 * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L
```

**Rationale**: Standard Etendo license headers typically use '(C)' instead of '©' and the '2021-2026' range for consistent character encoding and compliance with the core repository standards.

File: packages/MainUI/utils/expressions.ts [hunk 5]
```diff
@@ -216,6 +220,21 @@ export const createEvaluationContext = (options: SmartContextOptions) => {
       // parseDynamicExpression replaces OB.Utilities.getValue(obj, prop) with obj["prop"],
       // removing the null->'' conversion that getValue provided. The Proxy must handle it.
 
+      // Dev-mode warning: surface @field_name@ tokens that cannot be resolved so
+      // developers can spot misconfigured Application Dictionary expressions early.
+      // warnedRefs deduplicates so each missing key is logged at most once per context.
+      if (
+        process.env.NODE_ENV !== "production" &&
+        /^[a-zA-Z_]\w*$/.test(prop) &&
+        !["then", "toJSON", "toString", "valueOf", "constructor", "Symbol"].includes(prop) &&
```
Comment (etendobot) [inline packages/MainUI/utils/expressions.ts:229]: **Suggestion**

Move the identifier regex and the list of excluded properties outside the Proxy's 'get' trap.

```typescript
// Outside createEvaluationContext or the Proxy
const IDENTIFIER_REGEX = /^[a-zA-Z_]\w*$/;
const EXCLUDED_PROPS = ["then", "toJSON", "toString", "valueOf", "constructor", "Symbol"];

// Inside the trap
IDENTIFIER_REGEX.test(prop) &&
!EXCLUDED_PROPS.includes(prop) &&
```

**Rationale**: Defining regular expressions and arrays inside a frequently called function (like a Proxy get trap) causes unnecessary re-allocation. Moving them to a higher scope improves performance during expression evaluation.

File: packages/MainUI/utils/expressions.ts [hunk 6]
```diff
@@ -216,6 +220,21 @@ export const createEvaluationContext = (options: SmartContextOptions) => {
...
+        !warnedRefs.has(prop)
+      ) {
+        warnedRefs.add(prop);
+        console.warn(
+          `[DisplayLogic] Unresolvable reference "${prop}" — defaulting to "". Check that the field exists in the Application Dictionary and is included in the form context.`
+        );
+      }
+
       return defaultValue !== undefined ? defaultValue : "";
     },
```
Comment (etendobot) [inline packages/MainUI/utils/expressions.ts:239]: **Suggestion**

Use the nullish coalescing operator (??) to handle both null and undefined values.

```typescript
return defaultValue ?? "";
```

**Rationale**: The comment on line 217 mentions that the Proxy must handle the null->'' conversion previously provided by OB.Utilities.getValue. Using nullish coalescing (??) ensures that both null and undefined values are converted to an empty string, whereas the current implementation only handles undefined.

**Review by etendobot — APPROVED** [review]

File: packages/MainUI/utils/__tests__/expressions/dependencies.test.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    it("falls back to the raw token when DB column not found in fields", () => {
+      const deps = extractDependenciesFromExpression("@unknownField@='Y'", fields);
+      expect(deps).toEqual(["unknownField"]);
+    });
+  });
+
+  describe("Edge cases", () => {
+    it("returns empty array for undefined expression", () => {
+      expect(extractDependenciesFromExpression()).toEqual([]);
+    });
```
Comment (etendobot) [inline packages/MainUI/utils/__tests__/expressions/dependencies.test.ts:101]: **Suggestion**

Consider keeping the explicit 'undefined' argument to improve the clarity of the test intent.

```typescript
expect(extractDependenciesFromExpression(undefined)).toEqual([]);
```

**Rationale**: Explicitly passing 'undefined' in a test case labeled 'returns empty array for undefined expression' more clearly communicates the intent of the test. While calling the function without arguments typically results in an undefined parameter, being explicit helps document that this specific edge case is being handled.

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]
